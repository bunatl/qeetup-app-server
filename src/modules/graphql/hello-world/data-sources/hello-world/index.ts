import { TextDataSource } from '@container/data-sources/text';

export class HelloWorld extends TextDataSource {
    getText() {
        return Promise.resolve('Hello World!');
    }
}
