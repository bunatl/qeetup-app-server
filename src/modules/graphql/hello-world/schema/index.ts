import { gql } from 'apollo-server';

export const typeDefs = gql`
    extend type Query {
        test: String!
    }
`;
