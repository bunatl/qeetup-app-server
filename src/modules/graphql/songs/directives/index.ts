import { SongExistenceDirective } from './song-existence';

export const directives = {
    songExists: SongExistenceDirective,
};
