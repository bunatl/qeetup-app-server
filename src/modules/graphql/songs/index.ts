export { directives } from './directives';
export { resolvers } from './resolvers';
export { typeDefs } from './schema';
