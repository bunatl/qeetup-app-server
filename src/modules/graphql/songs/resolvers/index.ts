import { Resolvers } from '@container/resolvers';
import { queries } from './queries';

export const resolvers: Resolvers = {
    Query: queries,
};
