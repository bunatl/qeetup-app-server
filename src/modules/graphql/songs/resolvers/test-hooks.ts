import { gql } from 'apollo-server';

export const SONG_FRAGMENT = gql`
    fragment SongFragment on Song {
        name
        artist
        cover
        description
        listens
        tags {
            value
            isImportant
        }
        audio
        isLiked
        comments {
            user {
                name
                avatar
                isArtist
            }
            text
        }
    }
`;
