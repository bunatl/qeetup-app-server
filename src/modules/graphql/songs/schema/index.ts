import { gql } from 'apollo-server';

export const typeDefs = gql`
    extend type Query {
        songs: [Song!]!
    }

    type Song {
        name: String!
        artist: String!
        cover: String!
        description: String!
        listens: Int!
        tags: [Tag!]!
        audio: String
        isLiked: Boolean
        comments: [Comment!]
    }

    type Tag {
        value: String!
        isImportant: Boolean
    }

    type Comment {
        user: User!
        text: String!
    }

    type User {
        name: String!
        avatar: String!
        isArtist: Boolean
    }
`;
