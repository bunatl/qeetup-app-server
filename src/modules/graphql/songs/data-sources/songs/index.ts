import { InMemoryDataSource } from '@container/data-sources/in-memory';
import { Song } from '@container/schema';

const songs: Song[] = [
    {
        artist: 'September 87',
        name: 'Bad Dream Baby',
        listens: 145202,
        isLiked: true,
        cover: 'https://i.vimeocdn.com/video/526078435_1280x720.jpg',
        description: 'This song is by September 87 and features Dream Fiend.',
        tags: [
            {
                isImportant: true,
                value: 'New Synth Wave',
            },
            {
                isImportant: true,
                value: 'ðŸ”¥Popular',
            },
            {
                value: 'Electronic',
            },
            {
                value: 'Retro',
            },
            {
                value: 'Outrun',
            },
        ],
        comments: [
            {
                user: {
                    name: 'John Wick',
                    avatar:
                        'https://cdnb.artstation.com/p/assets/images/images/013/127/649/large/michael-papatsounis-johnwickfinal.jpg?1538170551',
                },
                text: 'This song is worth killing for',
            },
        ],
    },
    {
        name: 'Running in the night',
        artist: 'FM-84',
        listens: 32000,
        isLiked: false,
        cover: 'https://i.redd.it/xvyfcvt7fvwz.jpg',
        description: 'This song is by September 87 and features Dream Fiend.',
        tags: [
            {
                isImportant: true,
                value: 'New Synth Wave',
            },
            {
                isImportant: true,
                value: 'ðŸ”¥ Popular',
            },
            {
                value: 'Electronic',
            },
            {
                value: 'Retro',
            },
            {
                value: 'Outrun',
            },
        ],
    },
    {
        name: 'Kickstart my heart',
        artist: 'MÃ¶tley CrÃ¼e',
        listens: 1030002,
        isLiked: true,
        cover: 'https://townsquare.media/site/295/files/2019/02/MCscr.jpg?w=980&q=75',
        description: 'This song is by September 87 and features Dream Fiend.',
        tags: [
            {
                value: 'Glam metal',
            },
            {
                isImportant: true,
                value: 'ðŸ”¥ Popular',
            },
            {
                value: 'Heavy metal',
            },
            {
                value: 'Hard Rock',
            },
        ],
    },
    {
        name: 'Dark side of the moon',
        artist: 'Pink Floyd',
        cover: 'https://images-na.ssl-images-amazon.com/images/I/81aTawcGdmL._SL1500_.jpg',
        listens: 8398893,
        isLiked: false,
        description: 'Placeholder',
        tags: [
            {
                value: 'Progressive rock',
                isImportant: true,
            },
            {
                value: 'Rock',
            },
        ],
    },
    {
        name: 'Mind of a Beast',
        artist: 'The Glitch Mob',
        cover: 'https://consequenceofsound.net/wp-content/uploads/2014/01/91vaxjhtjjl-_sl1500_.jpg?quality=80&w=807',
        listens: 490000,
        isLiked: true,
        description: 'Placeholder',
        tags: [
            {
                value: 'Glitch',
                isImportant: true,
            },
            {
                value: 'Electronic rock',
            },
            {
                value: 'Dubstep',
            },
            {
                value: 'EDM',
            },
        ],
    },
];

export class SongsDataSource extends InMemoryDataSource {
    getAll() {
        return songs;
    }
}
