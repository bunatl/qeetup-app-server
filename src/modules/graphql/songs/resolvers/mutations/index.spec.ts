import { createApolloTestClient } from '@container/apollo/test-hooks';
import { CommentInput, Toggle } from '@container/schema';
import { SongsDataSource } from '@graphql/songs/data-sources';
import { gql } from 'apollo-server';

const SET_LIKE = gql`
    mutation SetLike($songId: ID!, $like: Toggle!) {
        setLike(songId: $songId, like: $like) {
            name
            artist
            cover
            description
            listens
            tags {
                value
                isImportant
            }
            audio
            isLiked
            comments {
                user {
                    name
                    avatar
                    isArtist
                }
                text
            }
        }
    }
`;

const ADD_COMMENT = gql`
    mutation AddComment($songId: ID!, $comment: CommentInput!) {
        addComment(songId: $songId, comment: $comment) {
            name
            artist
            cover
            description
            listens
            tags {
                value
                isImportant
            }
            audio
            isLiked
            comments {
                user {
                    name
                    avatar
                    isArtist
                }
                text
            }
        }
    }
`;

describe('Songs Mutations', () => {
    it('should return an error when song with given ID does not exist', async () => {
        expect.assertions(3);

        const { mutate } = createApolloTestClient();
        const { errors } = await mutate({
            mutation: SET_LIKE,
            variables: { songId: 'non-existing', like: Toggle.ADD },
        });

        expect(errors).not.toBeNull();
        expect(errors).toHaveLength(1);
        expect(errors![0].message).toMatch("Song with ID 'non-existing' does not exists.");
    });

    it('should like the song', async () => {
        expect.assertions(2);

        const songs = new SongsDataSource();
        const allSongs = await songs.getAll();
        const songId = allSongs[1].id;

        const { mutate } = createApolloTestClient();
        const { data: song } = await mutate({
            mutation: SET_LIKE,
            variables: { songId, like: Toggle.ADD },
        });

        expect(song!.setLike.isLiked).toBe(true);
        expect(song).toMatchSnapshot();
    });

    it('should unlike the song', async () => {
        expect.assertions(2);

        const songs = new SongsDataSource();
        const allSongs = await songs.getAll();
        const songId = allSongs[0].id;

        const { mutate } = createApolloTestClient();
        const { data: song } = await mutate({
            mutation: SET_LIKE,
            variables: { songId, like: Toggle.REMOVE },
        });

        expect(song!.setLike.isLiked).toBe(false);
        expect(song).toMatchSnapshot();
    });

    it('should add comment to the song', async () => {
        expect.assertions(2);

        const songs = new SongsDataSource();
        const allSongs = await songs.getAll();
        const songId = allSongs[1].id;

        const comment: CommentInput = {
            text: 'Test comment',
            user: {
                name: 'test',
                avatar: '...',
            },
        };

        const { mutate } = createApolloTestClient();
        const { data: song } = await mutate({
            mutation: ADD_COMMENT,
            variables: { songId, comment },
        });

        expect(song!.addComment.comments[0].text).toMatch(comment.text);
        expect(song).toMatchSnapshot();
    });
});
