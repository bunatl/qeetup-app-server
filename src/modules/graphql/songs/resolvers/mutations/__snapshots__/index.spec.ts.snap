// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Songs Mutations should add comment to the song 1`] = `
Object {
  "addComment": Object {
    "artist": "FM-84",
    "audio": null,
    "comments": Array [
      Object {
        "text": "Test comment",
        "user": Object {
          "avatar": "...",
          "isArtist": null,
          "name": "test",
        },
      },
    ],
    "cover": "https://i.redd.it/xvyfcvt7fvwz.jpg",
    "description": "This song is by September 87 and features Dream Fiend.",
    "isLiked": true,
    "listens": 32000,
    "name": "Running in the night",
    "tags": Array [
      Object {
        "isImportant": true,
        "value": "New Synth Wave",
      },
      Object {
        "isImportant": true,
        "value": "ðŸ”¥ Popular",
      },
      Object {
        "isImportant": null,
        "value": "Electronic",
      },
      Object {
        "isImportant": null,
        "value": "Retro",
      },
      Object {
        "isImportant": null,
        "value": "Outrun",
      },
    ],
  },
}
`;

exports[`Songs Mutations should like the song 1`] = `
Object {
  "setLike": Object {
    "artist": "FM-84",
    "audio": null,
    "comments": null,
    "cover": "https://i.redd.it/xvyfcvt7fvwz.jpg",
    "description": "This song is by September 87 and features Dream Fiend.",
    "isLiked": true,
    "listens": 32000,
    "name": "Running in the night",
    "tags": Array [
      Object {
        "isImportant": true,
        "value": "New Synth Wave",
      },
      Object {
        "isImportant": true,
        "value": "ðŸ”¥ Popular",
      },
      Object {
        "isImportant": null,
        "value": "Electronic",
      },
      Object {
        "isImportant": null,
        "value": "Retro",
      },
      Object {
        "isImportant": null,
        "value": "Outrun",
      },
    ],
  },
}
`;

exports[`Songs Mutations should unlike the song 1`] = `
Object {
  "setLike": Object {
    "artist": "September 87",
    "audio": null,
    "comments": Array [
      Object {
        "text": "This song is worth killing for",
        "user": Object {
          "avatar": "https://cdnb.artstation.com/p/assets/images/images/013/127/649/large/michael-papatsounis-johnwickfinal.jpg?1538170551",
          "isArtist": null,
          "name": "John Wick",
        },
      },
    ],
    "cover": "https://i.vimeocdn.com/video/526078435_1280x720.jpg",
    "description": "This song is by September 87 and features Dream Fiend.",
    "isLiked": false,
    "listens": 145202,
    "name": "Bad Dream Baby",
    "tags": Array [
      Object {
        "isImportant": true,
        "value": "New Synth Wave",
      },
      Object {
        "isImportant": true,
        "value": "ðŸ”¥ Popular",
      },
      Object {
        "isImportant": null,
        "value": "Electronic",
      },
      Object {
        "isImportant": null,
        "value": "Retro",
      },
      Object {
        "isImportant": null,
        "value": "Outrun",
      },
    ],
  },
}
`;
